### PR TITLE
[FAB-18485]Fabric orderer - add config option for client keepalives

### PR DESCRIPTION
**Pull request by Param-S**
added config option for orderer client keepalives.  Defaults defined as per the
existing hard coded values. 

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

### CODE REVIEW

File: orderer/common/localconfig/config_test.go [hunk 1]
```diff
@@ -314,3 +314,17 @@ func TestChannelParticipationDefaults(t *testing.T) {
...
+func TestKeepaliveDefaults(t *testing.T) {
+	os.Setenv("ORDERER_GENERAL_KEEPALIVE_CLIENTINTERVAL", "30s")
+	defer os.Unsetenv("ORDERER_GENERAL_KEEPALIVE_CLIENTINTERVAL")
+	cleanup := configtest.SetDevFabricConfigPath(t)
+	defer cleanup()
+
+	cc := &configCache{}
+	cfg, err := cc.load()
+	require.NotNil(t, cfg, "Could not load config")
+	require.NoError(t, err, "Load good config returned unexpected error")
```
Comment (C0rWin) [inline orderer/common/localconfig/config_test.go]:
```suggestion
	require.NoError(t, err, "Load good config returned unexpected error")
	require.NotNil(t, cfg, "Could not load config")
```

A better practice would be to first check for the error and only then for config not being nil.

Comment (Param-S) [reply orderer/common/localconfig/config_test.go]: sure. i will update the PR.

**Review by guoger — CHANGES_REQUESTED** [review]
I failed to see where newly added options are being used?

**Comment by Param-S** [comment]
Sorry, I missed out changes in clientconfig(Orderer/common/server/main.go) where these new options are consumed. I included it into PR now.

**Review by denyeart — COMMENTED** [review]
Also need to update orderer.yaml for the client settings.

The server settings are already there:
https://github.com/hyperledger/fabric/blob/main/sampleconfig/orderer.yaml#L43-L52

File: orderer/common/localconfig/config_test.go [hunk 2]
```diff
@@ -314,3 +314,17 @@ func TestChannelParticipationDefaults(t *testing.T) {
...
+	defer os.Unsetenv("ORDERER_GENERAL_KEEPALIVE_CLIENTINTERVAL")
+	cleanup := configtest.SetDevFabricConfigPath(t)
+	defer cleanup()
+
+	cc := &configCache{}
+	cfg, err := cc.load()
+	require.NoError(t, err, "Load good config returned unexpected error")
+	require.NotNil(t, cfg, "Could not load config")
+	require.Equal(t, 30*time.Second, cfg.General.Keepalive.ClientInterval)
+	require.Equal(t, Defaults.General.Keepalive.ServerMinInterval, cfg.General.Keepalive.ServerMinInterval)
```
[the remark this answers is not in the document]
Comment (Param-S) [reply orderer/common/localconfig/config_test.go:334]: Ok. I took the option of validating one config & one default in this case. I will include other options.

File: orderer/common/server/main.go [hunk 3]
```diff
@@ -532,9 +532,14 @@ func configureClusterListener(conf *localconfig.TopLevel, generalConf comm.Serve
 }
 
 func initializeClusterClientConfig(conf *localconfig.TopLevel) (comm.ClientConfig, bool) {
+	kaOpts := comm.DefaultKeepaliveOptions
+	kaOpts.ClientTimeout = conf.General.Keepalive.ClientTimeout
+	if conf.General.Keepalive.ClientInterval > time.Duration(0) {
+		kaOpts.ClientInterval = conf.General.Keepalive.ClientInterval
+	}
```
[the remark this answers is not in the document]
Comment (Param-S) [reply orderer/common/server/main.go]: Agree, I will make it consistent with server config.

**Comment by Param-S** [comment]
@guoger @denyeart updated the PR as per your comments. could you review the PR again.

**Comment by denyeart** [comment]
The code looks good to me now.
I thought I would take the opportunity to try it out and watch for the keepalives to verify it works as we would expect.
I added debug to the config code and I can see the ClientInterval is getting set as expected.

However, when using Wireshark, I don't see any keepalives from the follower (client) to leader (server) as I would expect.
I do however see keepalive messages from the server to client every 15 seconds, which is strange because I didn't set anything to 15 seconds, nor is 15 seconds a default in orderer or grpc go, as far as I can tell.

Another thing that is strange is that these keepalives every 15 seconds always seem to go from server to client, however when I restart an orderer, sometimes the 15 second keepalives go from follower to leader, and sometimes they go from leader to follower (but always from server side to client side).  This makes me think the connections are not getting established as Yacov described to us.

I couldn't figure out how to enable go grpc or http trace to see the connections and keepalives, which is why I'm using Wireshark.

@yacovm Any ideas on how to verify which grpc connections are established? Are there some debug messages I should look for, or would I need to add some more? Any clue why I would see a 15 second keepalive that I didn't configure?

**Comment by denyeart** [comment]
Here is an example of the 15s keepalive message in Wireshark, note that it goes from 6050 to 59746. My orderer is listening on 6050, so if I'm reading this correctly, it is a server to client keepalive message.

keepalives

**Comment by yacovm** [comment]
> I do however see keepalive messages from the server to client every 15 seconds, which is strange because I didn't set anything to 15 seconds, nor is 15 seconds a default in orderer or grpc go, as far as I can tell.

Could it be the leader heartbeat? Leaders send heartbeats to followers on a periodical basis. We don't know what is actually sent because in Raft it's encrypted with TLS. 

> however when I restart an orderer, sometimes the 15 second keepalives go from follower to leader, and sometimes they go from leader to follower (but always from server side to client side). 

Did you restart the leader by chance?

> Here is an example of the 15s keepalive message in Wireshark, note that it goes from 6050 to 59746. My orderer is listening on 6050, so if I'm reading this correctly, it is a server to client keepalive message. 

According to the [documentation](https://github.com/grpc/grpc-go/blob/master/Documentation/keepalive.md): "gRPC sends http2 pings on the transport to detect if the connection is down. " and [looking at the code](https://github.com/grpc/grpc-go/blob/v1.39.0/internal/transport/http2_client.go#L1514) it appears indeed to be the case. 
So you won't be able to see it as a keepalive in wireshark because it's encrypted. Even if it wasn't encrypted, I don't think wireshark classifies http/2 without you explicitly telling it to, and even then, probably it won't present it's an http/2 keepalive with such a label. 

Please next time upload the log so we can open it ourselves and understand better what is going on. 
As for debugging gRPC, can you perhaps try doing `GODEBUG=http2debug=2 ` ?

**Comment by denyeart** [comment]
Ok @yacovm I looked a bit deeper and I understand what is going on now.

The 15s server to client keepalive I'm seeing is for the deliver service (from orderer to peer), not the consensus service. If I stop the peer those messages go away. So we can ignore that.

Concerning consensus service, what I see is a leader (client) to follower (server port 6050) step message on connection1, followed immediately by a follower (client) to leader (server port 6051) response message on connection2. Here's the log snippet showing the step and response on different connections:

Leader (Raft2):
```
2021-07-15 14:29:20.163 EDT 01f1 DEBU [orderer.consensus.etcdraft] consensusSent -> Sending msg of 28 bytes to 1 on channel test-system-channel-name took 3.883µs
2021-07-15 14:29:20.163 EDT 01f3 DEBU [orderer.common.cluster.step] sendMessage -> Send of ConsensusRequest for channel test-system-channel-name with payload of size 28 to orderer.example.com(127.0.0.1:6050) took 41.735µs
2021-07-15 14:29:20.163 EDT 01f6 DEBU [orderer.common.cluster.step] handleMessage -> Received message from orderer.example.com(127.0.0.1:59225): ConsensusRequest for channel test-system-channel-name with payload of size 28
```

Follower (Raft1):
```
2021-07-15 14:29:20.163 EDT 0175 DEBU [orderer.common.cluster.step] handleMessage -> Received message from orderer2.example.com(127.0.0.1:59222): ConsensusRequest for channel test-system-channel-name with payload of size 28
2021-07-15 14:29:20.163 EDT 0176 DEBU [orderer.consensus.etcdraft] consensusSent -> Sending msg of 28 bytes to 2 on channel test-system-channel-name took 3.038µs
2021-07-15 14:29:20.163 EDT 0177 DEBU [orderer.common.cluster.step] sendMessage -> Send of ConsensusRequest for channel test-system-channel-name with payload of size 28 to orderer2.example.com(127.0.0.1:6051) took 68.247µs
```

So the response does not use the same connection1, it uses its own connection2. This is different than what you mentioned but I can accept it as fact since that is what the log shows.

Follower (client) to leader (server) transaction broadcast messages also uses the same connection2.

Since the default tick interval is 500ms, there will never be idle time on either connection. This is why I'm not seeing the keepalive messages that I had configured, they are truly not needed.  I then increased tick interval to 15s and set client keepalive to 12s, and now I do see the keepalive message 12 seconds after the step response.

So this proves that this PR to add a client keepalive configuration works as expected. But based on the default tick interval of 500ms and the fact that both connections are exercised for step messages, I think the conclusion is that we do not need a client keepalive after all (although it wouldn't hurt anything).

Does my assessment make sense to you? And if so, should we close this PR since no client keepalive is needed after all?

**Comment by mastersingh24** [comment]
@denyeart  - this makes sense.  I did not think that we actually had idle connections at the gRPC layer we actually do send ticks at the Raft layer.  I guess it does not hurt, but fundamentally I don't think it is actually needed.

**Comment by yacovm** [comment]
> @denyeart - this makes sense. I did not think that we actually had idle connections at the gRPC layer we actually do send ticks at the Raft layer. I guess it does not hurt, but fundamentally I don't think it is actually needed.

I think the rational behind this PR is that if you have a long lasting and idle connection from a follower to a leader, then a load balancer along the path might reap the connection, and then when a transaction will be forwarded from that follower to the leader, it will not get to the leader because the connection is dead.

**Comment by yacovm** [comment]
> @denyeart - this makes sense. I did not think that we actually had idle connections at the gRPC layer we actually do send ticks at the Raft layer. I guess it does not hurt, but fundamentally I don't think it is actually needed.

I **think** the rational behind this PR is that if you have a long lasting and idle connection from a follower to a leader, then a load balancer along the path might reap the connection, and then when a transaction will be forwarded from that follower to the leader, it will not get to the leader because the connection is dead. 

Unless you're saying that for every heartbeat from the leader, there is an ack from the follower? 

I think that you're right:

```
func (r *raft) handleHeartbeat(m pb.Message) {
	r.raftLog.commitTo(m.Commit)
	r.send(pb.Message{To: m.From, Type: pb.MsgHeartbeatResp, Context: m.Context})
}
```


> So the response does not use the same connection1, it uses its own connection2. This is different than what you mentioned but I can accept it as fact since that is what the log shows.


I never said they are on the same connection... I know that Raft nodes always send messages via client connections and receive them via server connections. I did write that code after all... 

> Follower (client) to leader (server) transaction broadcast messages also uses the same connection2.

We have a single TLS connection between every pair of nodes and all gRPC streams are multiplexed on it.


>  So this proves that this PR to add a client keepalive configuration works as expected. But based on the default tick interval of 500ms and the fact that both connections are exercised for step messages, I think the conclusion is that we do not need a client keepalive after all (although it wouldn't hurt anything).

How does it prove it? You are showing application logs and the gRPC keepalive is in a lower layer that is not quoted by you and I do not think it's even quoted.  

I don't see a reason to have a client keepalive if we do not need it. It has the potential to cause problems if misconfigured.

**Comment by denyeart** [comment]
@yacovm You are right, the reason we were looking into this is because we thought we had idle follower to leader connections that were getting reaped, and that aligned with your prior comments. But now we have proven that both connections get exercised at every step tick interval so the client keepalive is not needed.

How does it prove it? I used Wireshark to verify the step message use both connections every tick interval (aligned with the orderer logs), and to verify that client keepalive messages showed up if configured frequently enough using this PR (those do not show up in orderer logs but do show up in Wireshark as smaller messages).

Ultimately, I agree that this PR is not needed, as you say a misconfigured client keepalive will cause more problems than we were trying to solve (if configured too frequent the server will tell the client to go away). So let's close it if nobody else has comments.

**Comment by yacovm** [comment]
> How does it prove it? I used Wireshark to verify the step message use both connections every tick interval (aligned with the orderer logs), and to verify that client keepalive messages showed up if configured frequently enough using this PR (those do not show up in orderer logs but do show up in Wireshark as smaller messages).

Oh I see, you're saying that the when you set the keepalives to be more frequent than the heartbeats, the wireshark messages also increased in frequency accordingly.

**Comment by mastersingh24** [comment]
nice work @denyeart

**Comment by yacovm** [comment]
> Ultimately, I agree that this PR is not needed, as you say a misconfigured client keepalive will cause more problems than we were trying to solve (if configured too frequent the server will tell the client to go away). So let's close it if nobody else has comments.

Agreed, let Param close this.

**Pull request closed without merging**